### ISSUE TITLE
Add extras node types to `node-type.json`

### ISSUE THREAD

**Issue by CamilleTeruel** [tree-sitter/tree-sitter#617]
Thanks to the `node-type.json` file, we have type information that tells us which nodes we can encounter at each position in the parse trees of a given language. However, type information about nodes produced by `extras` rules is missing. With this information, `node-type.json` would give a complete description of parse trees, and client code would be able to know that those special node types can be found anywhere in a parse tree.
(Would this help AST generation in `haskell-tree-sitter`? @robrix)

This could be a represented with an `extras` flag, like the current `named`.

**Comment by patrickt** [comment]
cc @aymannadeem

**Comment by maxbrunsfeld** [comment]
This is a good idea, but I'm not sure how it would be structured. Do you mean that extra tokens would have an `extra` boolean flag in their top-level definition?

```js
{
  "type": "comment",
  "named": true,

  // NEW - this indicates that the `comment` node type can appear anywhere
  "extra": true 
}
```

**Comment by CamilleTeruel** [comment]
Yes, that's what I meant but I'm not sure what is best. 

Another solution could be to add a `"kind"` key to each top-level definition which can be `"extra"`, `"abstract_node"` or `"node"`. This also would make the distinction between abstract node types and concrete ones explicit (in our code we currently check for the presence of a `"subtypes"` key). OTOH, this assumes node "kinds" are mutually exclusive, I don't know if that conflicts with future features of tree-sitter (e.g. a node type that would be both abstract and extra, if that even make sense 😅).

**Comment by maxbrunsfeld** [comment]
I think I like the idea of just adding the `extra` flag.

**Comment by Razzeee** [comment]
Isn't this solved by https://github.com/tree-sitter/tree-sitter/pull/784 ?

**Comment by ahlinc** [comment]
> Isn't this solved by #784 ?

No, it wasn't, an additional marker for extras as Max suggested is not implemented yet.
